### PR TITLE
feat!: Integrate EMA Oracle

### PR DESCRIPTION
**Pull request by apopiak**
Integrate the EMA oracle from https://github.com/galacticcouncil/warehouse/pull/76 into the Basilisk runtime.

## Description

This PR adds the EMA oracle pallet and configures the runtime to ingest xyk trades and liquidity changes into it.
Supported periods are configured to `LastBlock`, `TenMinutes`, `Day` and `Week`.

The oracle values are not used, yet, to allow for observation.

## Related Issue

https://app.clickup.com/t/2v0w820

## Checklist:

- [ ] I have updated the documentation if necessary.
- [x] I have added tests to cover my changes, regression test if fixing an issue.
- [x] This is a breaking change.

### CODE REVIEW

File: pallets/exchange/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+[package]
+name = "pallet-exchange"
```
Comment (apopiak) [inline pallets/exchange/Cargo.toml:2]: confused that this is showing up here :thinking:

**Comment by apopiak** [comment]
```
[upgrade       ] 2023-01-24 13:43:46        REGISTRY: Unknown signed extensions CurrencyBalanceCheck found, treating them as no-effect
[upgrade       ] 2023-01-24 13:43:46        METADATA: Unknown types found, no types for Lookup74
[upgrade       ] connected to Testing Basilisk Local Testnet 12.0.1-4c240952f runtime 88.0.0 node 8.1.3 x86_64-linux-gnu
[upgrade       ] current runtime version 88
[upgrade       ] Error: createType(Call):: Call: failed decoding sudo.sudoUncheckedWeight:: Struct: failed on args: {"call":"Call","weight":"{\"refTime\":\"u64\"}"}:: Struct: failed on weight: {"refTime":"u64"}:: Struct: Cannot decode value 100 (typeof number), expected an input object, map or array
[upgrade       ]     at createTypeUnsafe (/home/runner/work/Basilisk-node/Basilisk-node/scripts/upgrade-runtime/node_modules/@polkadot/types/create/createType.cjs:75:18)
[upgrade       ]     at TypeRegistry.createTypeUnsafe (/home/runner/work/Basilisk-node/Basilisk-node/scripts/upgrade-runtime/node_modules/@polkadot/types/create/registry.cjs:417:45)
[upgrade       ]     at TypeRegistry.createType (/home/runner/work/Basilisk-node/Basilisk-node/scripts/upgrade-runtime/node_modules/@polkadot/types/create/registry.cjs:409:17)
[upgrade       ]     at extrinsicFn (/home/runner/work/Basilisk-node/Basilisk-node/scripts/upgrade-runtime/node_modules/@polkadot/types/metadata/decorate/extrinsics/createUnchecked.cjs:28:21)
[upgrade       ]     at Object.decorated [as sudoUncheckedWeight] (/home/runner/work/Basilisk-node/Basilisk-node/scripts/upgrade-runtime/node_modules/@polkadot/api/base/Decorate.cjs:462:22)
[upgrade       ]     at main (/home/runner/work/Basilisk-node/Basilisk-node/scripts/upgrade-runtime/index.js:69:28)
[upgrade       ]     at processTicksAndRejections (internal/process/task_queues.js:95:5)
[upgrade       ] npm ERR! code ELIFECYCLE
[upgrade       ] npm ERR! errno 1
[upgrade       ] npm ERR! upgrade-runtime@1.0.0 upgrade: `node index.js testing_basilisk_runtime.compact.compressed.wasm`
[upgrade       ] npm ERR! Exit status 1
[upgrade       ] npm ERR! 
[upgrade       ] npm ERR! Failed at the upgrade-runtime@1.0.0 upgrade script.
[upgrade       ] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[upgrade       ] 
[upgrade       ] npm ERR! A complete log of this run can be found in:
[upgrade       ] npm ERR!     /home/runner/.npm/_logs/2023-01-24T13_43_46_752Z-debug.log
```
The upgrade error doesn't seem related to this PR?

File: runtime/basilisk/src/lib.rs [hunk 2]
```diff
@@ -931,6 +932,23 @@ impl pallet_route_executor::Config for Runtime {
 	type WeightInfo = weights::route_executor::BasiliskWeight<Runtime>;
 }
 
+// constants need to be in scope to be used in generics
+use pallet_ema_oracle::{MAX_PERIODS, MAX_TRADES};
+
+parameter_types! {
+	pub SupportedPeriods: BoundedVec<OraclePeriod, ConstU32<MAX_PERIODS>> = BoundedVec::truncate_from(
+		vec![OraclePeriod::LastBlock, OraclePeriod::TenMinutes, OraclePeriod::Day, OraclePeriod::Week]
```
Comment (jak-pan) [inline runtime/basilisk/src/lib.rs:940]: keep LastBlock + DAY for now

File: pallets/exchange/Cargo.toml (hunk 1 above)
Comment (apopiak) [reply pallets/exchange/Cargo.toml:2]: math and warehouse seem fine, no? would prefer not to lose the history there
I can do this here, yeah

**Comment by apopiak** [comment]
superseded by #599

**Pull request closed without merging**